Objective-C/CBLDatabase.mm

```diff
@@ -627,18 +658,18 @@ - (nullable CBLDocument*) documentWithID: (NSString*)documentID
                                mustExist: (bool)mustExist
                                    error: (NSError**)outError
 {
-    [self mustBeOpen];
-    
-    return [[CBLDocument alloc] initWithDatabase: self
-                                      documentID: documentID
-                                       mustExist: mustExist
-                                           error: outError];
+    @synchronized (self) {
```

Comment:
I don't have a good solution now. But current thread-safe logic does not allow concurrent reads too. So it is better to address this as soon as possible.